### DISCUSSION TITLE
downloadArtifact data type

### DISCUSSION THREAD

**Discussion by pkrukp** [octokit/octokit.js#2040] [Q&A]
Hi,

I'm using nodejs and TypeScript.

`octokit.actions.downloadArtifact` returns response that has `data` property that has `unknown` type.

Is that normal (the `unknown` type) or am I doing something wrong? Debugger shows it's ArrayBuffer.

The problem is that I have to cast it manually to unzip it, manual casting is not nice :(

For example:
```
const downloadedArtifact = await octokit.actions.downloadArtifact({owner, repo, artifact_id, archive_format: "zip"});
await JSZip.loadAsync(downloadedArtifact.data as ArrayBuffer);
```

Thanks

**Comment by gr2m** [comment]
the response types are directly derived from GitHub's OpenAPI spec at https://github.com/github/rest-api-description/. We update our own version for @octokit at github.com/octokit/openapi where we maintain deprecations for the SDKs. From there we transpile the OpenAPI spec for api.github.com to TypeScript in https://github.com/octokit/openapi-types.ts. And finally we use that in https://github.com/octokit/types.ts/ where we maintain shared types for all the different @octokit modules.

In this case, the only response for that endpoint is

```yml
      responses:
        '302':
          description: response
          headers:
            Location:
              "$ref": "#/components/headers/location"
```

So we don't know what type it will have after following the redirect. I'm not sure if there is a solution to it. If it is always the same response type, we could add an octokit-specific amendment as part of `@octokit/openapi-types`, but I'm not sure if it's worth it. Only very few endpoints are affected

**Comment by pkrukp** [comment]
So what is the recommended approach? Check if it's ArrayBuffer and error otherwise?

```
const downloadedArtifact = await octokit.actions.downloadArtifact({owner, repo, artifact_id, archive_format: "zip"});
if (downloadedArtifact.data instanceof ArrayBuffer) {
   await JSZip.loadAsync(downloadedArtifact.data);
} else {
   // error
}
```

**Reply by gr2m** [reply]
we pass through the behavior from `node-fetch`. If the response type is not text of some kind, we return the promise of https://github.com/node-fetch/node-fetch#bodyarraybuffer. 

If you want full control over how the file is downloaded, e.g. if you want to stream the data, you can use your own request library

```js
const got = require("got")

const { url } = octokit.actions.downloadArtifact.endpoint({owner, repo, artifact_id, archive_format: "zip"});
const downloadStream = got.stream(url)
```

**Reply by pkrukp** [reply]
Thanks, I'd prefer to keep it simple. I'm using octokit for other endpoints so I'd prefer to do the same for downloadArtifact. I think checking for ArrayBuffer is quite safe so I'll just do that.
